### DISCUSSION TITLE
Track versions, stricter matching, and ignoring matches

### DISCUSSION THREAD

**Discussion by busywait** [music-assistant/support#1010] [Q&A]
I have a library of .flac files locally on Home Assistant in /media. They were ripped from CD with dpPowerAmp and include IRSC numbers where they were on the disc. Track IRSC numbers are missing from some CD rips, especially from the pre-1990s disks (IRSC numbers were first introduced in 1989 I think). All tracks have an Album name tag.

Sometimes I have a track on several versions of an album (eg track with Title "Kayleigh" is repeated with Album values "Misplaced Childhood", "Misplaced Childhood [1998 remaster]", and "Misplaced Childhood (2018 remix)". I also have versions of some of the tracks from live concert albums, eg the "The Thieving Magpie" album has live versions of all of the songs from Misplaced Childhood.

If I browse Music Assistant > Albums, search "Misplaced Childhood" and choose Play then I expect the non-remixed, non-remastered, non-live tracks that I selected to play, but I get a mix of studio and live versions of the songs. I also get crossfading between some tracks, even though I have "Strict" gapless playback turned on in the queue settings.

1/ Is it possible to stop these inappropriate matches by having stricter matching, perhaps of only tracks where the album name is the same?
2/ Is it possible to turn off Music Assistant track substitutions, because they can't offer me anything helpful in my local library?

**Comment by marcelveldt** [comment]
The matching is already very, very strict so if this happens the issue is probably with your tags.
Please check your tags if they have any issues. For example the version tag field is strictly matched, same applies to album etc.

So it can simply not happen that a live version is matched to a studio version etc.

**Reply by busywait** [reply]
My Album title tags look OK - all distinct. The Album Version seems to be extracted from that text (like "Remaster"), but seems to sometimes loose information like the 1998 in "1998 Remaster". My track Title tags often do not have any extra information appended, just the track name like "Kayleigh". Is that my problem?

Even so, this doesn't look correct according to what you described: 
![image](https://user-images.githubusercontent.com/3671165/205063864-4e7519fd-92cc-4bce-ac9b-c7bc87693375.png)

Is that as expected?

The tracks listed here have different Albums, why are they shown as "Linked"? And I have heard live versions in the played audio, the live albums all have very different album names to the studio albums.

Is there a way to turn the track substitution feature off while I'm trying to fix up my metadata?

**Reply by marcelveldt** [reply]
If there are the exact same track recording this is to be expected. The metadata will then contain an src that is an exact match.

**Reply by OzGav** [reply]
Just on your point about fixing the metadata the recommended way to do that is to use Musicbrainz Picard as it adds all of the tag information that MA uses.

**Reply by marcelveldt** [reply]
And no you can not disable track linking and I have no intention to add that to the code (as that is one of the reason what makes mA great) but instead work with you to sort out any issues.

What I can tell above you have exactly the same tracks on different albums hence they're linked.
If they are in fact different versions, they should have had a different src, recording id or version assigned in the tags

**Reply by busywait** [reply]
They are not the same track recording, but how would Music Assistant know? Made Again is a live album; the track version from Sympathy [Single] is a different live version, but Music Assistant is assuming they are the same version as the studio version. The Title for one of the tracks is "Kayleigh (live)", but the "(live)" part has been stripped out of the Title in the list in my previous post.

The tag information related to the tracks listed above: 
![image](https://user-images.githubusercontent.com/3671165/205066671-46ad4d8b-16bb-4cbd-af0b-3dbec54789a7.png)
![image](https://user-images.githubusercontent.com/3671165/205066963-1a5fe9b2-e225-4e50-af4e-ba75e6d09bc0.png)
![image](https://user-images.githubusercontent.com/3671165/205067376-9d738bfd-666c-4bea-b1fa-b6feb3a5521a.png)
![image](https://user-images.githubusercontent.com/3671165/205067552-bf4a7a67-8dc6-49d3-a8dc-65ccdcccb2ef.png)
![image](https://user-images.githubusercontent.com/3671165/205067978-dfadfbe5-1fed-442a-8ffa-585b45cab5d7.png)

I'll have a play with MusicBrainz Picard and see if it seems safe to let it start adding IDs to the files.

**Reply by marcelveldt** [reply]
Best results are with MusicBrainz. If you have your tags correct with Musicbrainz Picard, Music Assistant will 100% follow that. If tags are missing it tries guessing them (with various results), like parsing the "live" version from the title.

**Reply by busywait** [reply]
Thanks - I'll try that. The guessing algorithm clearly isn't working for my files 😄

**Reply by busywait** [reply]
Hi, I've been adding tags to some albums with MusicBrainz Picard. 
- Will I have to remove the db or reinstall MA to flush previous information from the db, because MA is still trying to link very different versions of tracks with similar names?

**Reply by busywait** [reply]
- Which code file tests what files are matches? I would like to be able to work out what the minimum tag changes that I will need to make are so that maybe I can automate the process a bit.

**Reply by busywait** [reply]
- If I have a perfectly complete set of Musicbrainz tags for one version of an album, but there is a track with a similar name (Title tag) in another folder that doesn't have Musicbrainz tags, will the tracks be linked? (So, do I need to have Musicbrainz IDs in every single track in my library?)

**Reply by OzGav** [reply]
Have a look here https://github.com/music-assistant/hass-music-assistant/discussions/543

**Reply by OzGav** [reply]
> Hi, I've been adding tags to some albums with MusicBrainz Picard.
> 
> * Will I have to remove the db or reinstall MA to flush previous information from the db, because MA is still trying to link very different versions of tracks with similar names?

You can do a force full sync via the menu top right or yes since you are just starting removing the db will also work but isn’t preferred

**Comment by OzGav** [comment]
Relates to v1
